bswap.1
.\"                              hey, Emacs:   -*- nroff -*-
.\" bswap is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program; see the file COPYING.  If not, write to
.\" the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
.\"
.TH BSWAP 1 "November 18, 2005"
.\" Please update the above date whenever this man page is modified.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins (default)
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
bswap \- Utility for bytes swapping, especially for endianess(little-/big-) converting.
.SH SYNOPSIS
.B bswap
.RI [ options ]
.SH DESCRIPTION
\fBbswap\fP washes your windows, mends your fences, mows your lawn...
.PP
It also...
.SH OPTIONS
\fBbswap\fP accepts the following options:
.TP
.BR  -o ", " --output=\fIname\fP
Write output to \fIname\fP instead of standard output.
.TP
.BR  -i ", " --interactive
Prompt for confirmation.
.TP
.B  --dry-run
Take no real actions.
.TP
.B  --no-warn
Disable warnings.
.TP
.BR  -q ", " --quiet ", " --silent
Inhibit usual output.
.TP
.B  --brief
Shorten output.
.TP
.BR  -v ", " --verbose
Print more information.
.TP
.BI  --directory= name
Use specified directory.
.TP
.BI  --cd= name
Change to specified directory before proceeding.
.TP
.BR  -? ", " --help
List all available options and their meanings.
.TP
.B  --usage
Show summary of options.
.TP
.BR  -V ", " --version
Show version of program.
.\" .SH "SEE ALSO"
.\" .BR foo (1), 
.\" .BR bar (1).
.SH AUTHOR
Meihui Fan <mhfan@ustc.edu>.


ChangeLog
2005-11-18  Meihui Fan  <mhfan@ustc.edu>

	* bswap: initial version.



bswap.c
/* 
   bswap - Utility for bytes swapping, especially for endianess(little-/big-) converting.

   Copyright (C) 2005 Meihui Fan

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2, or (at your option)
   any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software Foundation,
   Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.  

#include <termios.h>
#include <grp.h>
#include <pwd.h>
*/

#include <stdio.h>
#include <sys/types.h>
#include <argp.h>
#include "system.h"

#define EXIT_FAILURE 1

#if ENABLE_NLS
# include <libintl.h>
# define _(Text) gettext (Text)
#else
# define textdomain(Domain)
# define _(Text) Text
#endif
#define N_(Text) Text

char *xmalloc ();
char *xrealloc ();
char *xstrdup ();

static error_t parse_opt (int key, char *arg, struct argp_state *state);
static void show_version (FILE *stream, struct argp_state *state);

/* argp option keys */
enum {DUMMY_KEY=129
      ,BRIEF_KEY
      ,DRYRUN_KEY
      ,NOWARN_KEY
      ,CD_KEY
      ,DIRECTORY_KEY
};

/* Option flags and variables.  These are initialized in parse_opt.  */

char *oname;			/* --output=FILE */
FILE *ofile;
char *new_directory;		/* --cd=DIRECTORY */
char *desired_directory;	/* --directory=DIR */
int want_interactive;		/* --interactive */
int want_quiet;			/* --quiet, --silent */
int want_brief;			/* --brief */
int want_verbose;		/* --verbose */
int want_dry_run;		/* --dry-run */
int want_no_warn;		/* --no-warn */

static struct argp_option options[] =
{
  { "interactive", 'i',           NULL,            0,
    N_("Prompt for confirmation"), 0 },
  { "output",      'o',           N_("FILE"),      0,
    N_("Send output to FILE instead of standard output"), 0 },
  { "quiet",       'q',           NULL,            0,
    N_("Inhibit usual output"), 0 },
  { "silent",      0,             NULL,            OPTION_ALIAS,
    NULL, 0 },
  { "brief",       BRIEF_KEY,     NULL,            0,
    N_("Shorten output"), 0 },
  { "verbose",     'v',           NULL,            0,
    N_("Print more information"), 0 },
  { "dry-run",     DRYRUN_KEY,    NULL,            0,
    N_("Take no real actions"), 0 },
  { "no-warn",     NOWARN_KEY,    NULL,            0,
    N_("Disable warnings"), 0 },
  { "cd",          CD_KEY,        N_("DIRECTORY"), 0,
    N_("Change to DIRECTORY before proceeding"), 0 },
  { "directory",   DIRECTORY_KEY, N_("DIR"),       0,
    N_("Use directory DIR"), 0 },
  { NULL, 0, NULL, 0, NULL, 0 }
};

/* The argp functions examine these global variables.  */
const char *argp_program_bug_address = "<mhfan@ustc.edu>";
void (*argp_program_version_hook) (FILE *, struct argp_state *) = show_version;

static struct argp argp =
{
  options, parse_opt, N_("[FILE...]"),
  N_("Utility for bytes swapping, especially for endianess(little-/big-) converting."),
  NULL, NULL, NULL
};

int
main (int argc, char **argv)
{
  textdomain(PACKAGE);
  argp_parse(&argp, argc, argv, 0, NULL, NULL);

  /* TODO: do the work */

  exit (0);
}

/* Parse a single option.  */
static error_t
parse_opt (int key, char *arg, struct argp_state *state)
{
  switch (key)
    {
    case ARGP_KEY_INIT:
      /* Set up default values.  */
      oname = "stdout";
      ofile = stdout;
      new_directory = NULL;
      desired_directory = NULL;
      want_interactive = 0;
      want_quiet = 0;
      want_brief = 0;
      want_verbose = 0;
      want_dry_run = 0;
      want_no_warn = 0;
      break;

    case 'i':			/* --interactive */
      want_interactive = 1;
      break;
    case 'o':			/* --output */
      oname = xstrdup (arg);
      ofile = fopen (oname, "w");
      if (!ofile)
	argp_failure (state, EXIT_FAILURE, errno,
		      _("Cannot open %s for writing"), oname);
      break;
    case 'q':			/* --quiet, --silent */
      want_quiet = 1;
      break;
    case BRIEF_KEY:		/* --brief */
      want_brief = 1;
      break;
    case 'v':			/* --verbose */
      want_verbose = 1;
      break;
    case DRYRUN_KEY:		/* --dry-run */
      want_dry_run = 1;
      break;
    case NOWARN_KEY:		/* --no-warn */
      want_no_warn = 1;
      break;
    case CD_KEY:		/* --cd */
      new_directory = xstrdup (optarg);
      break;
    case DIRECTORY_KEY:		/* --directory */
      desired_directory = xstrdup (optarg);
      break;

    case ARGP_KEY_ARG:		/* [FILE]... */
      /* TODO: Do something with ARG, or remove this case and make
         main give argp_parse a non-NULL fifth argument.  */
      break;

    default:
      return ARGP_ERR_UNKNOWN;
    }
  return 0;
}

/* Show the version number and copyright information.  */
static void
show_version (FILE *stream, struct argp_state *state)
{
  (void) state;
  /* Print in small parts whose localizations can hopefully be copied
     from other programs.  */
  fputs(PACKAGE" "VERSION"\n", stream);
  fprintf(stream, _("Written by %s.\n\n"), "Meihui Fan");
  fprintf(stream, _("Copyright (C) %s %s\n"), "2005", "Meihui Fan");
  fputs(_("\
This program is free software; you may redistribute it under the terms of\n\
the GNU General Public License.  This program has absolutely no warranty.\n"),
	stream);
}
